http3/conn.go

```diff
@@ -0,0 +1,530 @@
+package http3
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/lucas-clemente/quic-go"
+	"github.com/lucas-clemente/quic-go/quicvarint"
+)
+
+const (
+	maxBufferedStreams   = 10
+	maxBufferedDatagrams = 10
+)
+
+// Conn is a base HTTP/3 connection.
+// Callers should use either ServerConn or ClientConn.
+type Conn interface {
+	// Settings returns the HTTP/3 settings for this side of the connection.
+	Settings() Settings
+
+	// PeerSettings returns the peer’s HTTP/3 settings.
+	// This will block until the peer’s settings have been received.
+	PeerSettings() (Settings, error)
+}
+
+// ServerConn is a server connection. It accepts and processes HTTP/3 request streams.
+type ServerConn interface {
+	Conn
+	AcceptRequestStream(context.Context) (RequestStream, error)
+}
+
+// ClientConn is a client connection. It opens and processes HTTP/3 request streams.
+type ClientConn interface {
+	Conn
+	OpenRequestStream(context.Context) (RequestStream, error)
+}
+
+// webTransportConn is an internal interface for implementing WebTransport.
+type webTransportConn interface {
+	acceptStream(context.Context, SessionID) (quic.Stream, error)
+	acceptUniStream(context.Context, SessionID) (quic.ReceiveStream, error)
+	openStream(SessionID) (quic.Stream, error)
+	openStreamSync(context.Context, SessionID) (quic.Stream, error)
+	openUniStream(SessionID) (quic.SendStream, error)
+	openUniStreamSync(context.Context, SessionID) (quic.SendStream, error)
+	readDatagram(context.Context, SessionID) ([]byte, error)
+	writeDatagram(SessionID, []byte) error
+}
+
+type connection struct {
+	session quic.EarlySession
+
+	settings Settings
+
+	peerSettingsDone chan struct{} // Closed when peer settings are read
+	peerSettings     Settings
+	peerSettingsErr  error
+
+	peerStreamsMutex sync.Mutex
+	peerStreams      [4]quic.ReceiveStream
+
+	incomingStreamsOnce    sync.Once
+	incomingStreamsErr     error
+	incomingRequestStreams chan RequestStream
+
+	incomingStreamsMutex sync.Mutex
+	incomingStreams      map[SessionID]chan quic.Stream // Lazily constructed
+
+	incomingUniStreamsOnce  sync.Once
```

Comment:
Would it make sense to start the go routine right away, and get rid of this `sync.Once`? Every HTTP/3 session will use unidirectional streams.

Reply:
True.

Deferring it simplifies the test mocking. My thought was that a Conn wouldn't start any goroutines  until asked to minimize resource consumption.